File: legacykit-ui/src/lib/api/apps.ts
import { invoke } from '@tauri-apps/api/core';

export type AppListScope = 'user' | 'system' | 'all';

export interface ListAppsRequest {
  scope: AppListScope;
}

export interface InstalledApp {
  bundleId: string;
  displayName: string | null;
  version: string | null;
}

export interface ListAppsResult {
  scope: AppListScope;
  apps: InstalledApp[];
}

export interface InstallIpaRequest {
  ipaPaths: string[];
}

export interface InstallIpaResult {
  installed: string[];
}

export interface UninstallAppRequest {
  bundleId: string;
}

export interface UninstallAppResult {
  bundleId: string;
}

export function listInstalledApps(request: ListAppsRequest): Promise<ListAppsResult> {
  return invoke<ListAppsResult>('list_installed_apps', { request });
}

export function installIpa(request: InstallIpaRequest): Promise<InstallIpaResult> {
  return invoke<InstallIpaResult>('install_ipa', { request });
}

export function uninstallApp(request: UninstallAppRequest): Promise<UninstallAppResult> {
  return invoke<UninstallAppResult>('uninstall_app', { request });
}

File: legacykit-ui/src-tauri/src/models/apps.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum AppListScope {
    User,
    System,
    All,
}

impl AppListScope {
    pub fn as_arg(&self) -> &'static str {
        match self {
            AppListScope::User => "--user",
            AppListScope::System => "--system",
            AppListScope::All => "--all",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ListAppsRequest {
    pub scope: AppListScope,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InstalledApp {
    pub bundle_id: String,
    pub display_name: Option<String>,
    pub version: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ListAppsResult {
    pub scope: AppListScope,
    pub apps: Vec<InstalledApp>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InstallIpaRequest {
    pub ipa_paths: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InstallIpaResult {
    pub installed: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UninstallAppRequest {
    pub bundle_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UninstallAppResult {
    pub bundle_id: String,
}
